src/statistics/Statistics.ts

```diff
@@ -0,0 +1,264 @@
+
+        this.logger.info('Statistics', 'Client statistics is enabled with period ' + periodSeconds + ' seconds.');
+    }
+
+    stop() {
+        if (this.intervalId != null) {
+            clearInterval(this.intervalId);
+        }
+    }
+
+    /**
+     * @param periodSeconds the interval at which the statistics collection and send is being run
+     */
+    schedulePeriodicStatisticsSendTask(periodSeconds: number): NodeJS.Timer {
+        return Util.scheduleWithRepetition(() => {
+            const ownerConnection: ClientConnection = this.getOwnerConnection();
+            if (ownerConnection == null) {
+                this.logger.trace('Statistics', 'Can not send client statistics to the server. No owner connection.');
+                return;
+            }
+
+            const stats: string[] = [];
+
+            this.fillMetrics(stats, ownerConnection);
+
+            this.addNearCacheStats(stats);
+
+            this.sendStats(stats.join(''), ownerConnection);
+        }, 0, periodSeconds * 1000);
+    }
+
+    sendStats(newStats: string, ownerConnection: ClientConnection) {
+        const request = ClientStatisticsCodec.encodeRequest(newStats);
+
+        this.client.getInvocationService().invokeOnTarget(request, ownerConnection.getAddress()).catch((err) => {
+            // suppress exception, do not print too many messages
+            this.logger.trace('Statistics', 'Could not send stats ', err);
+        });
+    }
+
+    /**
+     * @return the owner connection to the server for the client only if the server supports the client statistics feature
+     */
+    private getOwnerConnection(): ClientConnection {
+        const connection = this.client.getClusterService().getOwnerConnection();
+        if (connection == null) {
+            return null;
+        }
+
+        const ownerConnectionAddress: Address = connection.getAddress();
+        const serverVersion: number = connection.getConnectedServerVersion();
+        if (serverVersion < Statistics.FEATURE_SUPPORTED_SINCE_VERSION) {
+
+            // do not print too many logs if connected to an old version server
+            if (this.ownerAddress == null || !ownerConnectionAddress.equals(this.ownerAddress)) {
+                this.logger.trace('Statistics', 'Client statistics can not be sent to server '
+                    + ownerConnectionAddress + ' since, connected '
+                    + 'owner server version is less than the minimum supported server version ' +
+                    Statistics.FEATURE_SUPPORTED_SINCE_VERSION_STRING);
+
+            }
+            // cache the last connected server address for decreasing the log prints
+            this.ownerAddress = ownerConnectionAddress;
+            return null;
+        }
+        return connection;
+    }
+
+    private registerMetrics() {
+        this.registerGauge('os.committedVirtualMemorySize', () => Statistics.EMPTY_STAT_VALUE);
```

Comment:
Too many empty state gauges. registering them will mean unnecessary parsing. do we really need to register unsupplied stats?

Reply:
I don't know exact reason. I referenced the C# client. See: https://github.com/hazelcast/hazelcast-csharp-client/blob/89e7faa67ac5d25497d9f81bc43bcc7fcdf74442/Hazelcast.Net/Hazelcast.Client/Statistics.cs#L170